packages/mermaid/src/utils/imperativeState.ts

```diff
@@ -0,0 +1,37 @@
+export const createImperativeState = <S extends Record<string, unknown>>(init: () => S) => {
+  const state = init();
+
+  return {
+    get records() {
+      return state;
+    },
+    reset: () => {
+      Object.keys(state).forEach((key) => {
+        delete state[key];
+      });
+      Object.entries(init()).forEach(([key, value]: [keyof S, any]) => {
+        state[key] = value;
+      });
+    },
+  };
+};
+
+export const domain = {
+  optional: <V>(value?: V) => value,
+  identity: <V>(value: V) => value,
```

Comment:
This does not seem to be used outside the test. Move it into the test file?

Reply:
This functions are for db files written in typescript and will be useful in the future

Reply:
After the TypeScript to JavaScript compilation, these functions will still exist and slow down our code (although only by a tiny amount!).

It's potentially worth replacing them with `as <TYPE>` and `as <TYPE> | undefined`, but the syntax is not as nice and the performance impact isn't too big so I'm on the fence, e.g.

```typescript
state = createImperativeState(() => {
  foo: domain.optional<string>("default"),
});
// this is equivalent (but faster)
state = createImperativeState(() => {
  foo: "default" as string | undefined,
});
```

Reply:
I've replace cif with assert, but this way is not strict enough. For example, it's possible: `const value = {} as string;`. Constrained identity functions are commonly used in functional typescript and modern javascript engines can inline functions like this one.